csharp/ql/src/semmle/code/csharp/AnnotatedType.qll

```diff
@@ -77,24 +77,32 @@ private module Annotations {
   }
 
   newtype TAnnotations =
-    TAnnotationFlags(int flags) {
-      flags = getElementTypeFlags(_) or
-      flags = getTypeArgumentFlags(_, _) or
-      flags = getTypeParameterFlags(_, _)
+    TAnnotationFlags(int flags, Nullability n) {
+      exists(Element e | flags = getElementTypeFlags(e) and n = getElementNullability(e))
+      or
+      flags = 0 // n is unbound
     }
 
-  /** A set of annotations on a type. */
   class TypeAnnotations extends TAnnotations {
     int flags;
+    Nullability nullability;
 
-    TypeAnnotations() { this = TAnnotationFlags(flags) }
+    TypeAnnotations() { this = TAnnotationFlags(flags, nullability) }
 
-    /** Gets an annotation in this set of annotations. */
-    TypeAnnotation getAnAnnotation() { isSet(result.getBit()) }
+    /** Gets the flags (as a bitset) of this type annotation. */
+    int getFlags() { result = flags }
 
-    private predicate isSet(int bit) {
-      isBit(bit) and
-      exists(int mask | mask = getBitMask(bit) | flags.bitAnd(mask) = mask)
+    /** Gets the nullability of this type annotation. */
+    Nullability getNullability() { result = nullability }
+
+    /**
+     * Gets the `i`th "child" of this type annotation.
+     * This is used to represent structured datatypes, where the structure
+     * of the type annotation mirrors the structure of the annotated type.
+     */
+    bindingset[i]
```

Comment:
I don't much like exposing this predicate, as it doesn't bind `i`. Perhaps it should just be a private top-level predicate instead?